feat(sdk): add visual cues in sdk around evaluation usage and feature flags

heypoom opened the pull request:
Closes https://github.com/metabase/metabase/issues/47824

### Description

Adds a visual indicator for evaluation usage (e.g. using API keys in development) and feature flags

### How to verify

Follow the test cases listed in `enterprise/frontend/src/embedding-sdk/lib/usage-problem.ts` and in the test code at `enterprise/frontend/src/embedding-sdk/components/private/SdkUsageProblem/SdkUsageProblemDisplay.unit.spec.tsx`. This is also described in [this part of the product doc](https://www.notion.so/metabase/Add-visual-cues-in-SDK-around-evaluation-usage-and-feature-flag-004be55a2ed247839cc9521e9a58f6d4?pvs=4#52b68b5b877240339a1d38bfcf0d4e0f).

For example, you can use the SDK with API keys in localhost, and the warning indicator should pop up. If you click on it, the card should open.

### Demo

Testing each test cases for visual cues

<div>
    <a href="https://www.loom.com/share/117c21d3fd954d0194855f56370b33c6">
      <p>Testing each cases in the Visual Cues PR - Private Loom</p>
    </a>
    <a href="https://www.loom.com/share/117c21d3fd954d0194855f56370b33c6">
      <img style="max-width:300px;" src="https://cdn.loom.com/sessions/thumbnails/117c21d3fd954d0194855f56370b33c6-4da3dabb5e4544fa-full-play.gif">
    </a>
  </div>

Shows a warning when we are using API keys in development

![CleanShot 2567-09-11 at 05 17 46@2x](https://github.com/user-attachments/assets/4eea1f02-8470-4fa0-a5c9-c51470f946ad)

Shows an error when we are using API keys on production

![CleanShot 2567-09-11 at 05 48 15@2x](https://github.com/user-attachments/assets/a057efe4-5605-431f-a701-6300dc2892a4)

Renders properly in Shoppy's Pug theme

![CleanShot 2567-09-11 at 21 15 07@2x](https://github.com/user-attachments/assets/fb74246b-e554-4a50-8b6a-5efcd6fbd5cf)

Renders properly in Shoppy's Luminara theme

![CleanShot 2567-09-11 at 21 16 05@2x](https://github.com/user-attachments/assets/84d79970-e1b1-4fbd-b061-4e187701e955)

Renders properly in Shoppy's Stitch theme

![CleanShot 2567-09-11 at 21 15 52@2x](https://github.com/user-attachments/assets/77a392ad-a05e-461d-8be4-84e7ec49b07b)

### Caveat

There is a known edge case with CORS on non-localhost origins. If the user hasn't entered their license key and haven't configured their authorized origins yet, the CORS headers will be missing from the /api/session/properties endpoint, therefore all requests on non-localhost origins will fail. [We're addressing this.](https://metaboat.slack.com/archives/C063Q3F1HPF/p1726050178995979)

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

WiNloSt committed:
pass `embedding-sdk` flag in `token-features`

heypoom committed:
add license problem hook

heypoom committed:
add warning banner

heypoom committed:
style cue component

heypoom committed:
update cue component

heypoom committed:
report issue to console

heypoom committed:
report issue to console

heypoom committed:
use sdk loader for pre-initialization

heypoom committed:
extract print problem method

heypoom committed:
use non-standard shadows

heypoom committed:
add unit tests

heypoom committed:
prevent console log spam during tests

heypoom committed:
mock the embedding token feature

heypoom committed:
mock the embedding token feature by default

heypoom committed:
add more test cases

heypoom committed:
add test case

heypoom committed:
add test cases

heypoom committed:
increase popover z-index

heypoom commented on enterprise/frontend/src/embedding-sdk/components/private/AppInitializeController.tsx:
```diff
@@ -23,14 +26,33 @@ export const AppInitializeController = ({
+
+  const content = useMemo(() => {
+    if (!isInitialized) {
+      return <SdkLoader />;
```
I noticed we were using the `Loading...` text, so might be a good idea to use the loader component - this falls back to the loading indicator, which would look nicer than text.

heypoom commented on enterprise/frontend/src/embedding-sdk/components/private/AppInitializeController.tsx:
```diff
@@ -23,14 +26,33 @@ export const AppInitializeController = ({
+      return <SdkLoader />;
+    }
+
+    if (licenseProblem?.severity === "error") {
+      return null;
+    }
```
We make the components underneath the `MetabaseProvider` not render if it encountered a license / usage error, mainly to prevent misuse or abuse of SDK.

npretto replied:
If they put the provider as wrapper of the entire app, this will make the entire customer's app break if they have an issue with the license

heypoom replied:
@npretto fixed in commit https://github.com/metabase/metabase/pull/47821/commits/fd208075e24658a7718571f24d9d4ddc814cc616 where it only makes the SDK components not load, as opposed to the whole page. Thanks!

heypoom commented on enterprise/frontend/src/embedding-sdk/components/private/AppInitializeController.tsx:
```diff
@@ -23,14 +26,33 @@ export const AppInitializeController = ({
+      {content}
+
+      {licenseProblem && (
+        <Box pos="fixed" bottom="15px" left="15px">
```
The indicator floats on the bottom-left corner.

![CleanShot 2567-09-11 at 05 24 12@2x](https://github.com/user-attachments/assets/467cf432-c7d8-4dab-acb0-a5d04a12fd50)

heypoom committed:
simplify test setup

heypoom committed:
simplify test setup

heypoom commented on enterprise/frontend/src/embedding-sdk/components/private/SdkLicenseProblemBanner/SdkLicenseProblemBanner.module.css:
```diff
@@ -0,0 +1,34 @@
+  box-shadow: 0 6px 24px 0 #0000001a;
+}
+
+.Indicator.Warning {
+  border: 1px solid #e8b923;
+}
+
+.Indicator.Error {
+  border: 1px solid #f7a4a3;
+}
```
These colors doesn't have to be themed as it's not user-facing, and it's not the same as our CSS variables - the color differs a bit. I'll try to see if I can use `color-mix(...)` to derive these colors from our base palette.

heypoom commented on enterprise/frontend/src/embedding-sdk/components/private/SdkLicenseProblemBanner/SdkLicenseProblemBanner.module.css:
```diff
@@ -0,0 +1,34 @@
+  border-radius: 4px 0 0 4px;
+}
+
+.Indicator.Warning .Content {
+  background: #fef3cf;
+}
+
+.Indicator.Error .Content {
+  background: #fdf0f0;
+}
```
Same as the above. Will see if `color-mix` can get rid of hard-coding here.

heypoom commented on enterprise/frontend/src/embedding-sdk/components/private/SdkLicenseProblemBanner/SdkLicenseProblemBanner.stories.tsx:
```diff
@@ -0,0 +1,34 @@
+
+// eslint-disable-next-line import/no-default-export
+export default {
+  title: "EmbeddingSDK/SdkLicenseProblemBanner",
```
We don't have Loki tests configured for SDK yet. We can use `yarn storybook-embedding-sdk` in the meantime to make changes to this component easier.

![CleanShot 2567-09-11 at 05 33 35@2x](https://github.com/user-attachments/assets/4e9f2069-1074-4b77-a0d5-4bea4a291bb1)

heypoom commented on enterprise/frontend/src/embedding-sdk/components/private/SdkLicenseProblemBanner/SdkLicenseProblemBanner.tsx:
```diff
@@ -0,0 +1,110 @@
+            className={S.Content}
+          >
+            <Icon
+              name={isError ? "warning_round_filled" : "warning"}
```
The SVG is named "warning_round_filled" in Figma.

heypoom commented on enterprise/frontend/src/embedding-sdk/components/private/SdkLicenseProblemBanner/SdkLicenseProblemBanner.module.css:
```diff
@@ -0,0 +1,34 @@
+.PopoverDropdown.PopoverDropdown {
+  border: none;
+}
```
Allows us to get rid of the grey border that is on PopoverDropdown by default.

heypoom commented on enterprise/frontend/src/embedding-sdk/components/private/SdkLicenseProblemBanner/SdkLicenseProblemBanner.module.css:
```diff
@@ -0,0 +1,34 @@
+.Indicator {
+  cursor: pointer;
+  border-radius: 4px;
+  box-shadow: 0 6px 24px 0 #0000001a;
```
The shadow in the Figma design differs by what we have in our `Paper` component. This is fine as it needs to be visually distinctive on top of the user's application.

heypoom commented on enterprise/frontend/src/embedding-sdk/hooks/private/get-auth-configuration.ts:
```diff
@@ -52,10 +47,6 @@ export const getAuthConfiguration = (
       jwtProviderUri => setupJwtAuth(jwtProviderUri, dispatch),
     )
     .with([{ apiKey: P.select(P.nonNullable) }, "localhost"], apiKey => {
-      presentApiKeyUsageWarning(appName);
       setupLocalApiKey(dispatch, apiKey);
     })
-    .with([{ apiKey: P.select(P.nonNullable) }, P.not("localhost")], () =>
-      getErrorMessage("PROD_API_KEY"),
-    )
```
Both the logic for presenting the API key warnings and for showing errors when using API keys in production are part of the new code, so those are removed from `getAuthConfiguration`.

heypoom commented on enterprise/frontend/src/embedding-sdk/hooks/private/use-init-data.ts:
```diff
@@ -39,12 +37,14 @@ export const useInitData = ({ config }: InitDataLoaderParameters) => {
-      `Using Metabase Embedding SDK, version ${EMBEDDING_SDK_VERSION}`,
-    );
-  }, []);
+    if (allowConsoleLog) {
+      // eslint-disable-next-line no-console
+      console.log(
+        // eslint-disable-next-line no-literal-metabase-strings -- Not a user facing string
+        `Using Metabase Embedding SDK, version ${EMBEDDING_SDK_VERSION}`,
+      );
+    }
+  }, [allowConsoleLog]);
```
I'm getting a lot of console log spam when running `yarn jest` which is annoying. This allows us (and our customers) to disable console logs.

heypoom commented on enterprise/frontend/src/embedding-sdk/hooks/private/use-sdk-license-problem.ts:
```diff
@@ -0,0 +1,39 @@
+  const appName = useSelector(getApplicationName);
+
+  const hasFeatureFlag = useSelector(state => {
+    // When the settings endpoint has not been called, we assume that the feature is enabled.
+    if (!state.settings.values?.["token-features"]) {
+      return true;
+    }
```
I tried to use the existing `useHasTokenFeature` hook, but it throws an error because it is missing this check.

heypoom commented on enterprise/frontend/src/embedding-sdk/hooks/private/use-sdk-license-problem.ts:
```diff
@@ -0,0 +1,39 @@
+      return true;
+    }
+
+    // TODO: replace this with "embedding-sdk" once the token feature PR landed.
```
Waiting for the PR for the feature flags project.

heypoom commented on enterprise/frontend/src/embedding-sdk/lib/license-problem.ts:
```diff
@@ -0,0 +1,71 @@
+  const isApiKey = !!apiKey;
+  const isLocalhost = getIsLocalhost();
+
+  return match({ hasFeatureFlag, isSSO, isApiKey, isLocalhost })
```
This is defined [in the product docs](https://www.notion.so/metabase/Add-visual-cues-in-SDK-around-evaluation-usage-and-feature-flag-004be55a2ed247839cc9521e9a58f6d4?pvs=4#52b68b5b877240339a1d38bfcf0d4e0f).

heypoom commented on enterprise/frontend/src/embedding-sdk/lib/user-warnings.ts:
```diff
@@ -0,0 +1,9 @@
+import { t } from "ttag";
+
+const ERROR_MESSAGE_MAP = {
```
This is our existing errors, to keep as-is.

heypoom commented on enterprise/frontend/src/embedding-sdk/test/server-mocks/sdk-init.ts:
```diff
@@ -26,7 +26,8 @@ import { createMockState } from "metabase-types/store/mocks";
   currentUser = createMockUser(),
   settingValues = createMockSettings(),
-  tokenFeatures = createMockTokenFeatures(),
+  // TODO: change this to "embedding_sdk" once the token feature PR landed.
+  tokenFeatures = createMockTokenFeatures({ embedding: true }),
```
Without the token features, the SDK test will fail due to the license check in the SDK.

heypoom committed:
Merge branch 'master' into add-visual-cues-in-sdk

heypoom commented:
PS. There is a known edge case with CORS on non-localhost origins. If the user hasn't entered their license key and haven't configured their authorized origins yet, the CORS headers will be missing from the `/api/session/properties` endpoint, therefore all requests on non-localhost origins will fail. [We're addressing this.](https://metaboat.slack.com/archives/C063Q3F1HPF/p1726050178995979)

heypoom commented on enterprise/frontend/src/embedding-sdk/hooks/private/use-sdk-license-problem.ts:
```diff
@@ -0,0 +1,39 @@
+  const hasLoggedRef = useRef(false);
+  const appName = useSelector(getApplicationName);
+
+  const hasFeatureFlag = useSelector(state => {
```
There is a known edge case with CORS on non-localhost origins. If the user hasn't entered their license key and haven't configured their authorized origins yet, the CORS headers will be missing from the `/api/session/properties` endpoint, therefore we can't query whether the feature flag exists. [We're addressing this.](https://metaboat.slack.com/archives/C063Q3F1HPF/p1726050178995979)

npretto commented on enterprise/frontend/src/embedding-sdk/components/private/AppInitializeController.tsx:
```diff
@@ -1,13 +1,16 @@
-import type { ReactNode } from "react";
-import { t } from "ttag";
+import { type ReactNode, useMemo } from "react";
```
heads up that I'm removing this component in the style leak PR, so we'll have to figure out what to do

heypoom replied:
@npretto Sounds good, I'll update the PR once that happens. We can just create a dedicated component for this, like `SdkLicenseCheckWrapper` that focuses on license check and indicator.

heypoom replied:
@npretto fixed in https://github.com/metabase/metabase/pull/47821/commits/4b8eade1d50502339a379898469c8e9f27f40250, I moved it out to be its own component.

heypoom committed:
do not load individual component when a license error happens

heypoom committed:
move SdkLicenseProblemDisplay component out

heypoom commented on enterprise/frontend/src/embedding-sdk/components/private/SdkLicenseProblem/SdkLicenseProblemBanner.module.css:
```diff
@@ -0,0 +1,34 @@
+  box-shadow: 0 6px 24px 0 #0000001a;
+}
+
+.Indicator.Warning {
+  border: 1px solid #e8b923;
+}
+
+.Indicator.Error {
+  border: 1px solid #f7a4a3;
+}
```
These four doesn't need to be themeable as it's not user-facing, but I'll try to use `color-mix` here and see if I can replace these.

heypoom committed:
replace colors with color-mix

heypoom committed:
replace colors with color-mix

heypoom committed:
make the text unthemed to ensure legibility

WiNloSt commented:
It looks weird on shoppy, maybe we need to play around with CSS a bit more.
![image](https://github.com/user-attachments/assets/66171d5e-3a4c-4f86-9736-85283ab1fb99)

heypoom commented:
@WiNloSt good catch, let me try with portals and make it a bit more resilient to themes and external styles.

WiNloSt commented on enterprise/frontend/src/embedding-sdk/lib/license-problem.ts:
```diff
@@ -0,0 +1,71 @@
+      toError(PROBLEMS.CONFLICTING_AUTH_METHODS),
+    )
+    .with({ isSSO: true, hasFeatureFlag: true }, () => null)
+    .with({ isSSO: true, hasFeatureFlag: false }, () =>
```
This condition doesn't work for me. I think the reason is that the `SdkLicenseProblemDisplay` is nested under some components that require authentication. When the authentication fails in this case, since there's no feature flag, JWT won't work. It fails to render the error entirely.
![image](https://github.com/user-attachments/assets/09bf9ad1-42c5-468a-a89d-62b4471a2693)

heypoom replied:
@WiNloSt I think that is likely because of the CORS error in `GET /api/session/properties` as well, as that returns a CORS error in the screenshot, we do not have the token features, so `hasFeatureFlag` becomes `true`.

(We fallback to `true` here because we don't exactly know when the settings would be populated, so we don't show a flash of error)

```
  const hasFeatureFlag = useSelector(state => {
    // When the settings endpoint has not been called, we assume that the feature is enabled.
    if (!state.settings.values?.["token-features"]) {
      return true;
    }

    // TODO: replace this with "embedding-sdk" once the token feature PR landed.
    return getTokenFeature(state, "embedding");
  });
```

WiNloSt replied:
I think the logic should change then. If you can run this function when everything is ready e.g. the SDK is initialized, or `GET /api/session/properties` fails for example, then `hasFeatureFlag` should be `false` by default and is only set to `true` when we can read the settings.

having it as `true` by default makes it [not following the spec ](https://www.notion.so/metabase/Add-visual-cues-in-SDK-around-evaluation-usage-and-feature-flag-004be55a2ed247839cc9521e9a58f6d4?pvs=4#7c7e5a08cc0e4fea845163303600b11c)

heypoom committed:
hard-code colors to prevent them from being themed

heypoom committed:
hard-code colors to prevent them from being themed

heypoom committed:
update border styles

heypoom committed:
make the chevron unthemed

heypoom committed:
unset console text color

heypoom committed:
only set stroke on errors

heypoom committed:
optimistic query settings

heypoom committed:
document the sdk license condition

heypoom committed:
Merge branch 'master' into add-visual-cues-in-sdk

heypoom commented:
@WiNloSt I updated the theming so now it works with themed usages such as Shoppy. I've also added support for checking the settings (using `enable-embedding` for now), and added comments on the optimistic behaviour. Thanks for the feedback and testing!

heypoom committed:
Merge branch 'add-visual-cues-in-sdk' of github.com:metabase/metabase into add-visual-cues-in-sdk

heypoom committed:
remove redundant condition

heypoom committed:
add notes on api keys always being allowed on localhost

heypoom committed:
move the no auth method message to warning component

heypoom commented on enterprise/frontend/src/embedding-sdk/hooks/private/get-auth-configuration.ts:
```diff
@@ -52,10 +46,5 @@ export const getAuthConfiguration = (
       jwtProviderUri => setupJwtAuth(jwtProviderUri, dispatch),
     )
     .with([{ apiKey: P.select(P.nonNullable) }, "localhost"], apiKey => {
-      presentApiKeyUsageWarning(appName);
       setupLocalApiKey(dispatch, apiKey);
-    })
-    .with([{ apiKey: P.select(P.nonNullable) }, P.not("localhost")], () =>
-      getErrorMessage("PROD_API_KEY"),
-    )
-    .otherwise(() => getErrorMessage("NO_AUTH_PROVIDED"));
```
These 3 errors and warnings are moved to the `license-problem`

WiNloSt commented on enterprise/frontend/src/embedding-sdk/lib/license-problem.ts:
```diff
@@ -0,0 +1,90 @@
+        toWarning(PROBLEMS.API_KEYS_WITHOUT_LICENSE),
+      )
+      // We do not allow using API keys in production.
+      .with({ isApiKey: true, hasTokenFeature: true }, () =>
+        toError(PROBLEMS.API_KEYS_WITH_LICENSE),
+      )
+      .with({ isApiKey: true, hasTokenFeature: false }, () =>
+        toError(PROBLEMS.API_KEYS_WITHOUT_LICENSE),
```
This won't get displayed because we don't initialize the SDK on non localhost.

![image](https://github.com/user-attachments/assets/506c3785-ee46-4d4a-a93c-0f7557d37465)

https://github.com/metabase/metabase/blob/cffaf2750a6ca94edae26a96a438e8ebd5ac631c/enterprise/frontend/src/embedding-sdk/hooks/private/get-auth-configuration.ts#L48-L50

`setupLocalApiKey` would never run.

heypoom replied:
@WiNloSt I just noticed that, good call. I think it would be beneficial I'll re-run through the cases today and ensure they work. With that said, I think we should display this error - otherwise people might forget that they are using API keys, and confused while API keys doesn't work.

I think the problem is that `SdkLicenseProblemBanner` is within the `AppInitializeController`, so it's being prevented from being displayed. I'll see if I can make it work with the SDK not being initialized.

heypoom replied:
@WiNloSt I moved the `SdkLicenseProblemBanner` so now it does display properly.

![CleanShot 2567-09-13 at 00 52 33@2x](https://github.com/user-attachments/assets/d63c21ce-600d-41dd-bf64-8e1d50e0ade4)

WiNloSt commented on frontend/test/__support__/ui.tsx:
```diff
@@ -88,6 +88,12 @@ export function renderWithProviders(
+
+    // Enable the embedding feature by default in SDK tests, unless explicitly disabled.
+    // Without this, SDK components will not render due to missing token features.
+    if (!storeInitialState.settings && initialState.settings) {
```
Why don't we need this before this PR? Does it mean this PR modifies how our SDK works? Or does it mean we use a different setup in this PR?

heypoom replied:
@WiNloSt It's because we are using JWT authentication in tests without the token features (e.g. without "embedding"), so without the token feature the components would not render - because of the check in PublicComponentWrapper.

WiNloSt replied:
Should this be in each test's setup instead? In this case I imagine you only need to add setup proper `token-features` in this PR's unitest's setup. Then you can leave this util as is.

There are other unit tests that set up token features in unit tests for example
https://github.com/metabase/metabase/blob/2b0dc0b0c6c13e466f71b21f2d809720045633a2/frontend/src/metabase/admin/settings/app/components/SettingsEditor/tests/embedding/setup.tsx#L34-L37

heypoom replied:
@WiNloSt As `PublicComponentWrapper` is a part of every SDK component, I would have to modify every other SDK-related tests (e.g. `InteractiveQuestion`, `StaticQuestion`, `StaticDashboard`) and other tests that uses JWT authentication in SDK.

This only affects `mode: "sdk"` tests that does not have a `settings` object provided to it, so tests that provides their own `settings` in the initial state should still work. If this feels too implicit, I can add this to every SDK test instead.

WiNloSt replied:
You have a point, let's keep it like this.

WiNloSt reviewed (commented):
I feel like more cases when we're not running on localhost are not properly tested. Like I've mentioned, if some of the code part aren't going to be asserted, then we shouldn't have them.

heypoom commented on enterprise/frontend/src/embedding-sdk/components/private/PublicComponentWrapper/PublicComponentWrapper.tsx:
```diff
@@ -17,6 +20,7 @@ export const PublicComponentWrapper = React.forwardRef<
   PublicComponentWrapperProps
 >(function PublicComponentWrapper({ children, className, style }, ref) {
   const loginStatus = useSdkSelector(getLoginStatus);
+  const licenseProblem = useSdkSelector(getLicenseProblem);
```
@WiNloSt It's good that you raised this, as I was struggling to name this too. We have 2 ways of displaying errors:

- Authentication error: goes to the `SdkError` error component, so the user can control this
- License or misuse of authentication schemes (configuration problem): should show up in our Warning banner

I'm thinking "authentication" would be misleading here then. I was thinking about naming it "SdkUsageProblem", but that feels too generic to me...

WiNloSt replied:
sounds good to me, I was thinking something along the line, config problem as well, but SDK usage sounds more in line with the intent of the project.

heypoom replied:
Thanks! I'll rename `SdkLicenseProblem` to `SdkUsageProblem` then to be a bit more inclusive.

heypoom replied:
Fixed in https://github.com/metabase/metabase/pull/47821/commits/4d665d8374887a23e0b4b3fab282d4fe89cb128b, thanks!

heypoom committed:
rename sdk license problem to usage problem

heypoom committed:
add eslint rule for stories

heypoom commented on enterprise/frontend/src/embedding-sdk/components/private/SdkLicenseProblem/SdkLicenseProblemBanner.stories.tsx:
```diff
@@ -0,0 +1,34 @@
+
+import { SdkLicenseProblemBanner } from "./SdkLicenseProblemBanner";
+
+// eslint-disable-next-line import/no-default-export
```
Nice idea, fixed in https://github.com/metabase/metabase/pull/47821/commits/219726e52c047be482ca68a3c9bbe42f420d02b5

heypoom committed:
rename sdk license problem to usage problem

heypoom committed:
move sdk usage problem display out of the init controller

heypoom committed:
specify fallback font if instance is down or CORS error

heypoom committed:
add comments on conditions pending on the CORS PR

heypoom committed:
add notes on non-localhost checks

heypoom commented:
@WiNloSt Thanks for reviewing with care! I addressed your comments above.

> I feel like more cases when we're not running on localhost are not properly tested. Like I've mentioned, if some of the code part aren't going to be asserted, then we shouldn't have them.

Agree. I just recorded a [private Loom video](https://www.loom.com/share/117c21d3fd954d0194855f56370b33c6?sid=5252cacc-2785-49b7-9480-e45e77b33811) that tests every condition we have defined in `usage-problem.ts`, and removed ones that do not match assertions. That should cover it 😄

heypoom committed:
make test ids constant

heypoom committed:
add test case for neither provided

heypoom committed:
fix types

heypoom commented on enterprise/frontend/src/embedding-sdk/components/private/SdkUsageProblem/SdkUsageProblemBanner.tsx:
```diff
@@ -0,0 +1,144 @@
+  const { severity } = problem;
+  const isError = severity === "error";
+
+  // When the font family cannot be loaded from the MB instance,
+  // due to MB instance outage or missing CORS header,
+  // we fall back to the system font.
+  const fontFamily = `${theme.fontFamily}, sans-serif`;
```
If Nicolo's PR adds the `sans-serif` fallback system font, then I'll remove this. I'm adding this for now so the text has a readable fallback font when the MB instance is down, or missing a CORS header.

WiNloSt commented on enterprise/frontend/src/embedding-sdk/hooks/private/use-init-data.ts:
```diff
@@ -57,17 +57,8 @@ export const useInitData = ({ config }: InitDataLoaderParameters) => {
-      );
-    }
-  }, [appName, config, dispatch, loginStatus.status]);
+    getAuthConfiguration(config, dispatch);
```
We should rename this function then. It's not returning anything, and actually has side effects that configure the configuration, so the previous name was also a bit misleading.

```suggestion
    configureAuthConfiguration(config, dispatch);
```

heypoom replied:
Fixed in https://github.com/metabase/metabase/pull/47821/commits/60fba7f93be0c07cf5a1f1823abc41d15a0dc0c5, thanks!

WiNloSt commented on enterprise/frontend/src/embedding-sdk/hooks/private/use-sdk-usage-problem.ts:
```diff
@@ -0,0 +1,54 @@
+  const { allowConsoleLog = true } = config;
+
+  const hasLoggedRef = useRef(false);
+  const appName = useSelector(getApplicationName);
```
I'm not sure where we have specced this. This is weird.

https://github.com/metabase/metabase/blob/54825a17312ee32702ea596fd20e3bdce3fee878/enterprise/frontend/src/embedding-sdk/lib/print-usage-problem.ts#L24-L27

Imagine seeing `My super awesome app Embedding SDK for React. Please see the doc` and the doc still links to Metabase.

I really think here we have to say `Metabase Embedding SDK for React`

heypoom replied:
Fixed in https://github.com/metabase/metabase/pull/47821/commits/4aa503024e958329401a25c7776a30cf617895b1, thanks! The `appName` was part of the existing log, but it indeed doesn't make sense.

WiNloSt commented on enterprise/frontend/src/embedding-sdk/hooks/private/use-sdk-usage-problem.ts:
```diff
@@ -0,0 +1,54 @@
+      return true;
+    }
+
+    // TODO: replace this with "embedding-sdk" once the token feature PR landed.
```
as I've mentioned before, I think you should cherry-pick the token feature change and not wait for the feature flag PR, since we already update all token to have `embeddin-sdk`.

That feature flag feature is more about other settings and admin settings change.

heypoom replied:
@WiNloSt I noticed that the `embedding-sdk` flag is the only one with dash-case, while other flags are all underscore_case. Should we fix this?

```tsx
export const tokenFeatures = [
  "attached_dwh",
  "advanced_permissions",
  "audit_app",
  "cache_granular_controls",
  "disable_password_login",
  "content_verification",
  "embedding",
  "embedding-sdk",
  "hosting",
] as const;
```

WiNloSt replied:
Answered in our DM that the quoted code is outdated, it should be lowercase `embedding_sdk`

WiNloSt commented on enterprise/frontend/src/embedding-sdk/lib/usage-problem.ts:
```diff
@@ -0,0 +1,95 @@
+   * TODO: these checks for non-localhost environments are pending on
+   *       the "allowing CORS for /api/session/properties" PR to be merged
+   *
+   * 1: (isSSO: true, isEnabled: false) -> EMBEDDING_SDK_NOT_ENABLED
```
```suggestion
   * 1: (isSSO: true, isEnabled: false) -> PROBLEMS.EMBEDDING_SDK_NOT_ENABLED
```

WiNloSt reviewed (approved):
I think this is almost close to be mergeable now, please see if you could address my feedback as you see fit.

heypoom committed:
address code reviews

heypoom committed:
address code review

Co-authored-by: Mahatthana (Kelvin) Nomsawadi <me@bboykelvin.dev>

heypoom committed:
remove appName from console logs

heypoom commented on enterprise/frontend/src/embedding-sdk/components/private/SdkUsageProblem/SdkUsageProblemDisplay.unit.spec.tsx:
```diff
@@ -0,0 +1,159 @@
+
+describe("SdkUsageProblemDisplay", () => {
+  it("does not show an error when JWT is provided with a license", () => {
+    setup({ config: createMockJwtConfig(), tokenFeatureEnabled: true });
```
Fixed in https://github.com/metabase/metabase/pull/47821/commits/60fba7f93be0c07cf5a1f1823abc41d15a0dc0c5, thanks!

heypoom committed:
rename embedding-sdk to embedding_sdk

heypoom committed:
Merge branch 'master' into add-visual-cues-in-sdk

npretto commented on enterprise/frontend/src/embedding-sdk/components/private/SdkUsageProblem/SdkUsageProblemBanner.tsx:
```diff
@@ -0,0 +1,144 @@
+  const fontFamily = `${theme.fontFamily}, sans-serif`;
+
+  return (
+    <Popover position="top-start" opened={expanded} onChange={setExpanded}>
```
is "top-start" for the text direction?
Do we show the visual clue on the right when the language direction is right-to-left?

heypoom replied:
@npretto this is for the position of the popover itself, so it would place the popover at trigger's (flex)-start on the top of the trigger. The visual cue should still be on the left.

npretto replied:
I was wondering if it would display in the wrong place on RTL languages, but I tested and it's ok

npretto commented on enterprise/frontend/src/embedding-sdk/components/private/SdkUsageProblem/SdkUsageProblemDisplay.unit.spec.tsx:
```diff
@@ -0,0 +1,158 @@
+    sdk: createMockSdkState(),
+  });
+
+  setupEnterprisePlugins();
```
Should we have this, as we don't have the plugins in the actual sdk?

heypoom replied:
@npretto good point, I'll get rid of that.

heypoom replied:
Fixed in https://github.com/metabase/metabase/pull/47821/commits/7465abcb6b00e1f448e19d581c1a6022258b777c, thanks!

npretto commented on enterprise/frontend/src/embedding-sdk/lib/usage-problem.ts:
```diff
@@ -0,0 +1,95 @@
+}
+
+const PROBLEMS = {
+  API_KEYS_WITHOUT_LICENSE: t`The embedding SDK is using API keys. This is intended for evaluation purposes and works only on localhost. To use on other sites, a license is required and you need to implement SSO.`,
```
I don't think we should localize those messages, it would only make debugging and reporting issues harder

heypoom replied:
Good point. I'll remove the localizations then.

heypoom replied:
Fixed in https://github.com/metabase/metabase/pull/47821/commits/705ca68d684eb5c2f98b31bf4f72a99e8e7d233e, thanks!

npretto commented on frontend/src/metabase/ui/components/utils/Portal/index.ts:
```diff
@@ -0,0 +1,2 @@
+export { Portal } from "@mantine/core";
```
@ranquild would you be ok with adding Portal to `metabase/ui`?

ranquild replied:
I'm okay with this

heypoom committed:
add the embedding_sdk premium token feature

heypoom committed:
do not setup enterprise plugin in sdk test

heypoom committed:
do not localize visual cue messages

WiNloSt commented on enterprise/frontend/src/embedding-sdk/hooks/private/setup-sdk-auth.unit.spec.tsx:
```diff
@@ -9,18 +8,17 @@ import api from "metabase/lib/api";
-  const appName = "Metabase";
-  return getAuthConfiguration(config, dispatch, appName);
+
+  return setupSdkAuth(config, dispatch);
```
nit: I wouldn't return anything here since we don't expect the return value.

```suggestion
  setupSdkAuth(config, dispatch);
```

heypoom replied:
Fixed in https://github.com/metabase/metabase/pull/47821/commits/d94cac934b6a6556e6bc05cc97ccc054d03beba1 and fixed the unit test, thanks!

heypoom committed:
fix unit test in setup-sdk-auth

heypoom committed:
rename tests

heypoom committed:
remove unused eslint-disable

heypoom committed:
enable embedding_sdk premium feature for sdk unit tests

heypoom committed:
Merge branch 'master' into add-visual-cues-in-sdk

heypoom committed:
revert security_test.clj to use embedding instead of embedding-sdk

heypoom committed:
remove unneeded test

heypoom committed:
remove unneeded test

heypoom committed:
changes to support the style leak fixes

heypoom committed:
Merge branch 'master' into add-visual-cues-in-sdk
